### PR TITLE
Allow Custom fields

### PR DESCRIPTION
**Pull request by elliott-pd**
Add Custom fields implementation.
Will throw ArgumentException if a reserved property is used e.g. Select or Filter.

### CODE REVIEW

File: src/ConnectSdk/Querying/ReservedQueryKeys.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace ConnectSdk.Querying
+{
+    internal class ReservedQueryKeys
+    {
+        public static string Collection = "collection";
```
Comment (AdamPD) [inline src/ConnectSdk/Querying/ReservedQueryKeys.cs]: I'm not loving this - looks like we're going to have to keep a big list of these at all times.  Is there any other way?

File: src/ConnectSdk/Querying/Query.cs [hunk 2]
```diff
@@ -50,21 +55,53 @@ public Query(string collection, IEventEndpoint endpoint, AliasedAggregations agg
...
+        }
+
         public override string ToString()
         {
-            return JsonConvert.SerializeObject(this, Formatting.None, Serializer);
+            if (_data == null)
+                return JsonConvert.SerializeObject(this, Formatting.None, Serializer);
+
+            // Create a new temporary dictionary so we don't actually put our other values into the CustomValue dictionary
+            var tempDictionary = _data.ToDictionary(pair => pair.Key, pair => pair.Value);
```
Comment (AdamPD) [inline src/ConnectSdk/Querying/Query.cs]: I don't really like this - before it was so much cleaner.  How can we keep it the old way?

File: src/ConnectSdk/Querying/Query.cs [hunk 3]
```diff
@@ -50,21 +55,53 @@ public Query(string collection, IEventEndpoint endpoint, AliasedAggregations agg
...
 
         public virtual async Task<QueryResponse<TResult>> Execute()
         {
              return await Endpoint.Query(Collection, this);
         }
-        
+
+        private void OverrideReservedValue(Dictionary<string, object> dict, string key, object value)
+        {
+            if (value == null)
```
Comment (AdamPD) [inline src/ConnectSdk/Querying/Query.cs]: If we did go with this - an extension method would be much better.  I'm still not loving it though.

**Review by AdamPD — CHANGES_REQUESTED** [review]

File: src/ConnectSdk/Querying/Query.cs (hunk 2 above)
Comment (elliott-pd) [reply src/ConnectSdk/Querying/Query.cs]: We can create an ExpandoObject (or a dynamic object) and apply each of the key value pairs to it, then apply our properties over the top and have that converted to a string. This does still have the issue that if we did want to add more properties to the Query class we have to remember to add them to this object in the ToString.

The issue comes down to how the JsonConverter handles our object. If we want each of the key value pairs to be appended to the Query object as properties we cannot do that dynamically as it's a typed class. 

If we can merge two JSON objects taking the properties from a second one we could convert the dictionary to a JSON object, then the apply the Query object over the top and return the result. That way we wouldn't have to keep track of individual properties, and we could drop the ReservedQueryKeys class, I do like that class though as we can then throw exceptions based on reserved key words. Preventing the user from trying to go `.Custom("select", {})` etc

**Review by AdamPD — APPROVED** [review]
Much nicer.

**Pull request merged**